src/WebJobs.Script/Environment/EnvironmentExtensions.cs

```diff
@@ -46,6 +46,12 @@ public static bool IsEasyAuthEnabled(this IEnvironment environment)
             return isEasyAuthEnabled;
         }
 
+        public static bool IsStaticWebAppsFunction(this IEnvironment environment)
+        {
+            bool.TryParse(environment.GetEnvironmentVariable(StaticWebAppsFunction), out bool isStaticWebAppFunction);
```

Comment:
In the meeting last week we discussed using a header from the MSHA as well. Have we settled on a solution that works for BYOF as well? @miwebst

Ideally, it'd be nice if we could do this just by looking at the shape of the JSON in `x-ms-client-principal` header. But if not, perhaps an extra SWA header might be more robust as the MSHA or local emulator can set it.

Reply:
Tagging @cgillum and @ConnorMcMahon in this conversation as well because I've had separate threads with them about this.

There are basically two things going on here @anthonychu that contribute to this design:
1. The content of the x-ms-client-principal header value is different for Easy Auth and Static Web Apps. This could, like you suggested be solved by looking at the shape of the json object in the header and using that to determine which of the two environments the header is used for. We could also address this by perhaps considering making a breaking change to make static web apps better line up with Easy Auth. This point alone does not require the presence of the environment variable.
2. The second piece is the fact that currently, the claims principal binding is only enabled when Easy Auth is enabled. This is to prevent spoofing via the manipulation of the x-ms-client-principal header values. Static Web Apps functions don't have Easy Auth enabled, so we need another way for the function to know that this function is a Static Web Apps function, so the requests are coming from the MSHA, so there isn't any spoofing happening on the x-ms-client-principal header values. That's where I have introduced this new environment variable STATIC_WEB_APPS_FUNCTION, which has to be set to true in order for the claims principal binding to be enabled. At the meeting last week, we discussed that instead of using a environment variable, let's have the MSHA add a header to the function request instead. However, if we go on with that approach, if the same function app is used in a non-static-web-app context, then that app is vulnerable to spoofing attacks by the requesting party setting the header the way the MSHA would and impersonating any victim by setting the x-ms-client-principal value as they wish. This is an edge case, but one that the security team wants us to consider. Chris did mention that there could be some work-arounds we could do to do something other than an environment variable, but this solution cannot be as straightforward as the MSHA adding some basic headers, for security reasons.

Reply:
Thanks for the explanation. That makes total sense why we need to rely on an environment variable.

Now I’m thinking through what’s needed to support this besides the changes in these PRs.
- When SWA function apps are created, the platform has to set the variable. This variable needs to be retroactively applied to existing SWA apps. Also need to add to list of variables blocked from user modification.
- In BYOF (which admittedly I don’t know much about), the customer needs to add it. Do we only ask that they set it if it’s a .net app and they need SWA auth? Could be hard to document and easy to make mistakes.
- Locally, the tooling probably should ensure this is set in local.settings.json in a SWA function app. Or perhaps we require that they use an emulator, which can set the variable when it spawns the core tools process.

To be completely transparent, given what I know now and after thinking through this, I wonder if we should consider another option which is to do nothing for now. Would love your thoughts on my assumptions above as well as whether this is the right feature to focus on right now.

Reply:
I agree that it'd be ideal if you could avoid the env var requirement. Regarding the spoofing concern - are requests originating from MSH AppService internal requests, direct to the worker, bypassing the FE? If so, you could use our existing [IsAppServiceInternalRequest](https://github.com/Azure/azure-functions-host/blob/1ea19bf594ec14294667e96173036d1a5ef0c365/src/WebJobs.Script/Extensions/HttpRequestExtensions.cs#L79) mechanism to determine that the request is coming from MSHA, and assuming MSHA validates/replaces any specific headers applied by the user to prevent spoofing, we could trust them here.

Reply:
@mathewc, I don't think they are internal requests, and with the down the road features like BYO Function apps and private MSHAs, I don't think we can rely on the fact that requests won't be coming through the front end.